contracts/tf20/src/contract.rs

```diff
@@ -0,0 +1,510 @@
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        // admin functions for the contract to control the tokenfactory denom
+        ExecuteMsg::Mint { recipient, amount } => {
+            Ok(mint(deps, env, info, recipient, amount)?)
+        }
+        ExecuteMsg::ForceTransfer { owner, recipient, amount } => {
+            Ok(force_transfer(deps, env, info, owner, recipient, amount)?)
+        }
+        ExecuteMsg::ForceBurn { owner , amount } => {
+            Ok(force_burn(deps, env, info, owner, amount)?)
+        }
+        ExecuteMsg::ForceSend { owner, contract, amount, msg} => {
+            Ok(force_send(deps, env, info, owner, contract, amount, msg)?)
+        }
+        ExecuteMsg::UpdateContractAdmin { new_admin } => {
+            Ok(update_contract_admin(deps, env, info, new_admin)?)
+        }
+        ExecuteMsg::UpdateTokenFactoryAdmin { new_admin } => {
+            Ok(update_tokenfactory_admin(deps, env, info, new_admin)?)
+        }
+        ExecuteMsg::ModifyMetadata { metadata } => {
+            Ok(modify_metadata(deps, env, info, metadata)?)
+        }
+
+        // these all come from cw20-base to implement the cw20 standard
+        ExecuteMsg::Transfer { recipient, amount } => {
+            Ok(transfer(deps, env, info, recipient, amount)?)
+        }
+        ExecuteMsg::Burn { amount } => Ok(burn(deps, env, info, amount)?),
+        ExecuteMsg::Send {
+            contract,
+            amount,
+            msg,
+        } => Ok(send(deps, env, info, contract, amount, msg)?),
+        ExecuteMsg::IncreaseAllowance {
+            spender,
+            amount,
+            expires,
+        } => Ok(execute_increase_allowance(
+            deps, env, info, spender, amount, expires,
+        )?),
+        ExecuteMsg::DecreaseAllowance {
+            spender,
+            amount,
+            expires,
+        } => Ok(execute_decrease_allowance(
+            deps, env, info, spender, amount, expires,
+        )?),
+        ExecuteMsg::TransferFrom {
+            owner,
+            recipient,
+            amount,
+        } => Ok(transfer_from(
+            deps, env, info, owner, recipient, amount,
+        )?),
+        ExecuteMsg::BurnFrom { owner, amount } => {
+            Ok(burn_from(deps, env, info, owner, amount)?)
+        }
+        ExecuteMsg::SendFrom {
+            owner,
+            contract,
+            amount,
+            msg,
+        } => Ok(send_from(
+            deps, env, info, owner, contract, amount, msg,
+        )?),
+    }
+}
+
+pub fn assert_admin(deps: Deps, sender: Addr) -> ContractResult<()> {
+    let token_info = TOKEN_INFO.load(deps.storage)?;
+    match token_info.admin {
+        None => Err(ContractError::Unauthorized),
+        Some(admin) => {
+            if sender != admin {
+                return Err(ContractError::Unauthorized);
+            }
+            Ok(())
+        }
+    }
+}
```

Comment:
We basically derive the token name from the creator and denom, but maybe we should just store the whole string

Reply:
I have switched to just storing the denom.

It is possible to query the tf admin directly, but that wouldn't work because the contract needs to be the token admin, otherwise it would not be able to do force transfers based on incoming msgs